### PR TITLE
feat: setup instructor dashboard filter

### PR DESCRIPTION
**Pull request by Ian2012**
### Description

This PR adds a model for email preferences, an API to update that record, and the integration to the instructor dashboard to handle the preferences:

image

### Testing instructions
In your environment, configure the following settings. The following plugin is used as reference:

```yaml
name: settings
version: 0.1.0
patches:
    openedx-common-settings: |
      OPEN_EDX_FILTERS_CONFIG = {
        "org.openedx.learning.instructor.dashboard.render.started.v1": {
          "fail_silently": False,
          "pipeline": [
            "platform_plugin_forum_email_notifier.extensions.filters.AddInstructorNotifierTab",
          ]
        },
      }
```
- Go to the instructor dashboard
- Go to Forum notification tab
- Select some of the available options
- Go to the admin and search for an stored record associated with that user, and that course for the preference you choose http://local.overhang.io:8000/admin/platform_plugin_forum_email_notifier

**Merge checklist:**
Check off if complete *or* not applicable:
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Fixup commits are squashed away
- [ ] Unit tests added/updated
- [ ] Manual testing instructions provided
- [ ] Noted any: Concerns, dependencies, migration issues, deadlines, tickets

### CODE REVIEW

**Comment by Ian2012** [comment]
@mariajgrimaldi how can I get the current user in the filter?

I need to inject the current preference of the user

**Comment by mariajgrimaldi** [comment]
@Ian2012: can't you get the user from the current request? You can also inspect the instructor template context, but I'm 90% sure it doesn't include the logged in user.

File: platform_plugin_forum_email_notifier/admin.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from django.contrib import admin
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/admin.py:2]: don't forget the module docstring

File: platform_plugin_forum_email_notifier/apps.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 """
 
 from django.apps import AppConfig
+from openedx.core.constants import COURSE_ID_PATTERN
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/apps.py]: We'd probably need to add a try-except

**Comment by mariajgrimaldi** [comment]
@Ian2012: can you add the test instructions? Thanks!

File: platform_plugin_forum_email_notifier/apps.py (hunk 2 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/apps.py]: Fixed

File: platform_plugin_forum_email_notifier/admin.py (hunk 1 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/admin.py:2]: fixed

**Comment by Ian2012** [comment]
@mariajgrimaldi updated

File: platform_plugin_forum_email_notifier/models.py [hunk 3]
```diff
@@ -1,24 +1,53 @@
...
+from django.utils.translation import gettext as _
 from model_utils.models import TimeStampedModel
+from opaque_keys.edx.django.models import CourseKeyField
+
+User = get_user_model()
+
+
+class PreferenceOptions(models.IntegerChoices):
+    """
+    A model to store forum email notification preferences for a user.
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/models.py:15]: Not exactly. What about `Preference options for a user related to email notifications frequency`?

File: platform_plugin_forum_email_notifier/tasks.py [hunk 4]
```diff
@@ -63,3 +65,45 @@ def send_email_notification(
...
+                user=user, course_id=course_id
+            )
+            if preference.preference in (
+                PreferenceOptions.NONE,
+                PreferenceOptions.ALL_POSTS_DAILY_DIGEST,
+                PreferenceOptions.ALL_POSTS_WEEKLY_DIGEST,
+            ):
+                continue
+        except User.DoesNotExist:
+            log.warning(f"User {subscriber} does not exist")
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/tasks.py:100]: we stop sending notifications if a user doesn't exist?

File: platform_plugin_forum_email_notifier/tasks.py [hunk 5]
```diff
@@ -63,3 +65,45 @@ def send_email_notification(
...
+    Get the subscribers for a thread and notify them.
+    """
+    if object_type == ForumObject.THREAD:
+        subscribers = get_subscribers(thread_id)
+    elif object_type in (ForumObject.RESPONSE, ForumObject.COMMENT):
+        subscribers = get_subscribers(discussion.get("id"))
+    else:
+        raise ValueError(f"Invalid thread event type: {object_type}")
+
+    staff_subscribers = get_staff_subscribers(course_id)
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/tasks.py:83]: what's the difference between getting subscribers and staff subscribers?

File: platform_plugin_forum_email_notifier/tasks.py (hunk 4 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/tasks.py:100]: Yes, this shouldn't happen, but if it happens the task will finish succesfully

File: platform_plugin_forum_email_notifier/tasks.py (hunk 5 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/tasks.py:83]: A subscriber is a user(staff or course staff) who is subscribed to a thread.

An staff subscriber is a course staff member that configured it's preference for the course to get notified of all posts

File: platform_plugin_forum_email_notifier/tasks.py (hunk 4 above)
Comment (mariajgrimaldi) [reply platform_plugin_forum_email_notifier/tasks.py:100]: why don't we ignore it and continue?

Comment (mariajgrimaldi) [reply platform_plugin_forum_email_notifier/tasks.py:100]: what if a preference option does not exist for the user?

File: platform_plugin_forum_email_notifier/tasks.py [hunk 6]
```diff
@@ -63,3 +65,45 @@ def send_email_notification(
...
+            user = User.objects.get(id=subscriber)
+            preference = ForumNotificationPreference.objects.get(
+                user=user, course_id=course_id
+            )
+            if preference.preference in (
+                PreferenceOptions.NONE,
+                PreferenceOptions.ALL_POSTS_DAILY_DIGEST,
+                PreferenceOptions.ALL_POSTS_WEEKLY_DIGEST,
+            ):
+                continue
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/tasks.py:98]: if we're filtering for ALL_POSTS for staff subscribers, why are checking for these preferences as well? do non-staff users have preferences?

File: platform_plugin_forum_email_notifier/utils.py [hunk 7]
```diff
@@ -32,6 +32,26 @@ def get_subscribers(thread_id):
     return subscribers
 
 
+def get_staff_subscribers(course_id):
+    """Return a list of user ids subscribed to a thread."""
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/utils.py]: can we be a bit more specific of what a staff subscriber is here in the docstring?

File: platform_plugin_forum_email_notifier/utils.py [hunk 8]
```diff
@@ -32,6 +32,26 @@ def get_subscribers(thread_id):
...
+
+    return subscribers
+
+
 def _url_for_thread_subscriptions(thread_id):
     """Return the url for the thread subscriptions endpoint in the forum service."""
     return f"{settings.PREFIX}/threads/{thread_id}/subscriptions"
+
+
+class ForumObject(IntEnum):
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/utils.py:53]: can we add a docstring here

**Comment by mariajgrimaldi** [comment]
Last few comments. Can we make the tests pass?

File: platform_plugin_forum_email_notifier/tasks.py (hunk 6 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/tasks.py:98]: Not at the moment, we should do that only for staff users

File: platform_plugin_forum_email_notifier/tasks.py (hunk 4 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/tasks.py:100]: that's will cause a failure. I'll fix it

File: platform_plugin_forum_email_notifier/tasks.py (hunk 6 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/tasks.py:98]: This is fixed in the email digest branch

File: platform_plugin_forum_email_notifier/tasks.py (hunk 4 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/tasks.py:100]: This is fixed in the email_digest branch

**Review by mariajgrimaldi — APPROVED** [review]
LGTM! We tested this in a remote environment :)!

**Comment by BryanttV** [comment]
Tested locally and it works!

**Pull request merged**